Mathlib/Analysis/SpecialFunctions/Pow/Real.lean

```diff
@@ -905,56 +905,94 @@ section Tactics
+    IsNat (a ^ b) ne := by
+  rwa [pb.out, Real.rpow_int_cast]
+
+theorem isInt_rpow_pos {a b : ℝ} {nb ne : ℕ}
+    (pb : IsNat b nb) (pe' : IsInt (a ^ nb) (Int.negOfNat ne)) :
+    IsInt (a ^ b) (Int.negOfNat ne) := by
+  rwa [pb.out, rpow_nat_cast]
+
+theorem isInt_rpow_neg {a b : ℝ} {nb ne : ℕ}
+    (pb : IsInt b (Int.negOfNat nb)) (pe' : IsInt (a ^ (Int.negOfNat nb)) (Int.negOfNat ne)) :
+    IsInt (a ^ b) (Int.negOfNat ne) := by
+  rwa [pb.out, Real.rpow_int_cast]
+
+theorem isRat_rpow_pos {a b : ℝ} {nb : ℕ}
+    {num : ℤ} {den : ℕ}
+    (pb : IsNat b nb) (pe' : IsRat (a^nb) num den) :
+    IsRat (a^b) num den := by
+  rwa [pb.out, rpow_nat_cast]
+
+theorem isRat_rpow_neg {a b : ℝ} {nb : ℕ}
+    {num : ℤ} {den : ℕ}
+    (pb : IsInt b (Int.negOfNat nb)) (pe' : IsRat (a^(Int.negOfNat nb)) num den) :
+    IsRat (a^b) num den := by
+  rwa [pb.out, Real.rpow_int_cast]
+
+/-- Evaluates expressions of the form `a ^ b` when `a` and `b` are both reals.-/
+@[norm_num (_ : ℝ) ^ (_ : ℝ)]
+def evalRPow : NormNumExt where eval {u α} e := do
+  let .app (.app _ (a : Q(ℝ))) (b : Q(ℝ)) ← Lean.Meta.whnfR e | failure
+  let sα ← inferDivisionRing α
+  haveI' : u =QL 0 := ⟨⟩
+  haveI' : $α =Q ℝ := ⟨⟩
+  haveI' : $e =Q $a ^ $b := ⟨⟩
+  let (rb : Result b) ← derive (α := q(ℝ)) b
+  match rb with
+  | .isBool .. | .isRat _ .. => failure
+  | .isNat sβ nb pb =>
+    let e' : Q(ℝ) := q($a ^ $nb)
+    match ← derive e' with
+    | .isBool .. => failure
+    | .isNat sα' ne' pe' =>
+      assumeInstancesCommute
+      haveI' : $sα' =Q AddGroupWithOne.toAddMonoidWithOne := ⟨⟩
+      return .isNat sα' ne' q(isNat_rpow_pos $pb $pe')
+    | .isNegNat sα' ne' pe' =>
+      assumeInstancesCommute
+      return .isNegNat sα' ne' q(isInt_rpow_pos $pb $pe')
+    | .isRat sα' qe' nume' dene' pe' =>
+      assumeInstancesCommute
+      have pe' : Q(@IsRat ℝ instRingReal («$a» ^ «$nb») «$nume'» «$dene'») := pe'
+      have pe'' := q(isRat_rpow_pos $pb $pe')
+      return .isRat sα' qe' nume' dene' pe''
+  | .isNegNat sβ nb pb =>
+    let e' : Q(ℝ) := q(($a ^ (-($nb : ℤ))))
+    match ← derive e' with
+    | .isBool .. => failure
+    | .isNat sα' ne' pe' =>
+      assumeInstancesCommute
+      return .isNat sα' ne' q(isNat_rpow_neg $pb $pe')
+    | .isNegNat sα' ne' pe' =>
+      assumeInstancesCommute
+      have pe' : Q(@IsInt ℝ instRingReal («$a» ^ Int.negOfNat «$nb») (Int.negOfNat «$ne'»)) := pe'
+      have pb : Q(@IsInt ℝ instRingReal «$b» (Int.negOfNat «$nb»)) := pb
+      have pe'' := q(isInt_rpow_neg $pb $pe')
+      return .isNegNat sα' ne' pe''
+    | .isRat sα' qe' nume' dene' pe' =>
+      assumeInstancesCommute
+      have pe' : Q(IsRat (@HPow.hPow ℝ ℤ ℝ
+                            (@instHPow ℝ ℤ
+                               (@DivInvMonoid.Pow ℝ
+                                  (@DivisionRing.toDivInvMonoid ℝ instDivisionRingReal))) «$a»
+                                    (Int.negOfNat «$nb»))
+                          $nume' $dene') := pe'
```

Comment:
Why the cursed expression here?

Reply:
Not sure. I'm just trying to give `Qq` what it's asking for.

Reply:
I pushed a fix; `assertInstancesCommute` requires you to first name the instance you want.